tsdb/chunks/chunk_write_queue.go

```diff
@@ -0,0 +1,180 @@
+// Adding jobs to the job is non-blocking as long as the queue isn't full.
+type chunkWriteQueue struct {
+	size int
+	jobs chan chunkWriteJob
+
+	chunkRefMapMtx       sync.RWMutex
+	chunkRefMap          map[ChunkDiskMapperRef]chunkenc.Chunk
+	chunkRefMapOversized bool // indicates whether more than <size> chunks were put into the chunkRefMap.
+
+	isRunningMtx sync.RWMutex // protects the isRunning property.
+	isRunning    bool         // used to prevent that new jobs get added to the queue when the chan is already closed.
+
+	workerWg sync.WaitGroup
+
+	writeChunk writeChunkF
+
+	operationsMetric *prometheus.CounterVec
+}
+
+// writeChunkF is a function which writes chunks, it is dynamic to allow mocking in tests.
+type writeChunkF func(HeadSeriesRef, int64, int64, chunkenc.Chunk, ChunkDiskMapperRef, bool) error
+
+func newChunkWriteQueue(reg prometheus.Registerer, size int, writeChunk writeChunkF) *chunkWriteQueue {
+	q := &chunkWriteQueue{
+		size:        size,
+		jobs:        make(chan chunkWriteJob, size),
+		chunkRefMap: make(map[ChunkDiskMapperRef]chunkenc.Chunk, size),
+		writeChunk:  writeChunk,
+
+		operationsMetric: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "prometheus_tsdb_chunk_write_queue_operations_total",
+				Help: "Number of operations on the chunk_write_queue.",
+			},
+			[]string{"operation"},
+		),
+	}
+
+	if reg != nil {
+		reg.MustRegister(q.operationsMetric)
+
+		// Initialize series for all the possible labels.
+		for _, op := range queueOperations {
+			q.operationsMetric.WithLabelValues(op).Add(0)
+		}
+	}
+
+	q.start()
+	return q
+}
+
+func (c *chunkWriteQueue) start() {
+	c.workerWg.Add(1)
+	go func() {
+		defer c.workerWg.Done()
+
+		for job := range c.jobs {
+			c.processJob(job)
+		}
+	}()
+
+	c.isRunningMtx.Lock()
+	c.isRunning = true
+	c.isRunningMtx.Unlock()
+}
+
+func (c *chunkWriteQueue) processJob(job chunkWriteJob) {
+	err := c.writeChunk(job.seriesRef, job.mint, job.maxt, job.chk, job.ref, job.cutFile)
+	if job.callback != nil {
+		job.callback(err)
+	}
+
+	c.chunkRefMapMtx.Lock()
+	defer c.chunkRefMapMtx.Unlock()
+
+	delete(c.chunkRefMap, job.ref)
+
+	if len(c.chunkRefMap) == 0 {
+		// If the map had to be grown beyond its allocated size, then we recreate it to free memory.
+		if c.chunkRefMapOversized {
+			c.chunkRefMap = make(map[ChunkDiskMapperRef]chunkenc.Chunk, c.size)
+			c.chunkRefMapOversized = false
+		}
+	}
+
+	c.operationsMetric.WithLabelValues(queueOperationComplete).Inc()
+}
+
+func (c *chunkWriteQueue) addJob(job chunkWriteJob) error {
+	c.isRunningMtx.RLock()
+	defer c.isRunningMtx.RUnlock()
+
+	if !c.isRunning {
+		return errors.New("queue is not started")
+	}
+
+	c.chunkRefMapMtx.Lock()
+	// The map might grow beyond the allocated size here, in which case we'll recreate it as soon as it is drained.
+	c.chunkRefMap[job.ref] = job.chk
+	if len(c.chunkRefMap) > c.size {
+		c.chunkRefMapOversized = true
+	}
```

Comment:
I was thinking about this and I think you can remove it. `len(c.chunkRefMap)` will be always less or equal than `c.size + 1` as you can only have `c.size` jobs in `c.jobs` plus the one inflight in the worker. 

Having this map allocated to one extra element doesn't seem like a memory leaking issue, and if it was, you can just initialise the `c.jobs` channel to `size-1`.

Reply:
@replay pointed me out offline that the mutex above is actually a RWMutex and we could still get more than `size+1` elements in that case. I suggested that maybe we should just switch that mutex to a normal mutex, we had a long conversation on this and decided to let the benchmark talk instead of guessing numbers.

<details>
<summary>Benchmark code</summary>

```go
func BenchmarkChunkWriteQueue_addJob(b *testing.B) {
	for _, concurrentWrites := range []int{1, 10, 100, 1000} {
		b.Run(fmt.Sprintf("%d concurrent writes", concurrentWrites), func(b *testing.B) {
			issueReadSignal := make(chan struct{}, b.N)
			q := newChunkWriteQueue(nil, 1000, func(ref HeadSeriesRef, i int64, i2 int64, chunk chunkenc.Chunk, ref2 ChunkDiskMapperRef, b bool) error {
				select {
				case issueReadSignal <- struct{}{}:
				default:
					// can't, don't block
				}
				return nil
			})
			b.Cleanup(func() {
				// stopped already, so no more writes will happen
				close(issueReadSignal)
			})
			b.Cleanup(q.stop)

			start := sync.WaitGroup{}
			start.Add(1)

			jobs := make(chan chunkWriteJob, b.N)
			for i := 0; i < b.N; i++ {
				jobs <- chunkWriteJob{
					seriesRef: HeadSeriesRef(i),
					ref:       ChunkDiskMapperRef(i),
				}
			}
			close(jobs)

			go func() {
				for range issueReadSignal {
					// we don't care about the ID we're getting, we just want to grab the lock
					_ = q.get(ChunkDiskMapperRef(0))
				}
			}()

			done := sync.WaitGroup{}
			done.Add(concurrentWrites)
			for w := 0; w < concurrentWrites; w++ {
				go func() {
					start.Wait()
					for j := range jobs {
						_ = q.addJob(j)
					}
					done.Done()
				}()
			}

			b.ResetTimer()
			start.Done()
			done.Wait()
		})
	}
}
```
</details>

That is an extreme case benchmark that tries to issue as many reads as writes, while writing concurrently, and the results are good:

```
name                                              old time/op  new time/op  delta
ChunkWriteQueue_addJob/1_concurrent_writes-16      822ns ± 2%   810ns ± 1%   -1.46%  (p=0.000 n=10+10)
ChunkWriteQueue_addJob/10_concurrent_writes-16     871ns ± 0%   836ns ± 2%   -4.01%  (p=0.000 n=10+10)
ChunkWriteQueue_addJob/100_concurrent_writes-16    980ns ± 2%   849ns ± 3%  -13.38%  (p=0.000 n=10+10)
ChunkWriteQueue_addJob/1000_concurrent_writes-16  1.08µs ± 2%  0.90µs ± 3%  -16.50%  (p=0.000 n=10+10)
```

I also tried benchmarking without the `get` calls, and with only writes the new (normal sync.Mutex, no map cleaning) is even better as the parallelism increases:

```
name                                              old time/op  new time/op  delta
ChunkWriteQueue_addJob/1_concurrent_writes         267ns ± 1%   248ns ± 1%   -6.96%  (p=0.000 n=8+9)
ChunkWriteQueue_addJob/1_concurrent_writes-4       272ns ± 1%   276ns ± 1%   +1.78%  (p=0.000 n=10+10)
ChunkWriteQueue_addJob/1_concurrent_writes-16      271ns ± 1%   275ns ± 1%   +1.29%  (p=0.000 n=10+10)
ChunkWriteQueue_addJob/10_concurrent_writes        284ns ± 3%   259ns ± 4%   -8.84%  (p=0.000 n=10+10)
ChunkWriteQueue_addJob/10_concurrent_writes-4      402ns ± 2%   289ns ± 1%  -28.22%  (p=0.000 n=10+10)
ChunkWriteQueue_addJob/10_concurrent_writes-16     396ns ± 1%   287ns ± 1%  -27.66%  (p=0.000 n=8+10)
ChunkWriteQueue_addJob/100_concurrent_writes       279ns ± 2%   259ns ± 2%   -7.10%  (p=0.000 n=10+10)
ChunkWriteQueue_addJob/100_concurrent_writes-4     501ns ± 0%   299ns ± 1%  -40.39%  (p=0.000 n=8+9)
ChunkWriteQueue_addJob/100_concurrent_writes-16    477ns ± 2%   298ns ± 1%  -37.62%  (p=0.000 n=10+10)
ChunkWriteQueue_addJob/1000_concurrent_writes      284ns ± 2%   266ns ± 2%   -6.33%  (p=0.000 n=10+10)
ChunkWriteQueue_addJob/1000_concurrent_writes-4    581ns ± 1%   308ns ± 1%  -46.91%  (p=0.000 n=8+9)
ChunkWriteQueue_addJob/1000_concurrent_writes-16   541ns ± 2%   310ns ± 1%  -42.66%  (p=0.000 n=10+9)
```

Reply:
Amazing that you spotted this, I didn't consider the performance of read locks versus normal locks.
Updated: de725f363d6482000a5e17e3069b31da3a253d72

Reply:
Now that we have the normal lock, we can get rid of the map-cleaning part, because now it's always bounded.

Reply:
I love how simple this becomes: https://github.com/prometheus/prometheus/pull/10051/commits/450e9bd908659af4e9aa9639ab1d47d9b82bbd12

Reply:
What do you think about incrementing the counter outside the lock scope? or am I making this unnecessarily complicated? 855f156aa

Reply:
that's the result of your benchmark with/without 855f56
```
benchmark                                                    old ns/op     new ns/op     delta 
BenchmarkChunkWriteQueue_addJob/1_concurrent_writes-4        797           808           +1.38%
BenchmarkChunkWriteQueue_addJob/10_concurrent_writes-4       856           864           +0.90%
BenchmarkChunkWriteQueue_addJob/100_concurrent_writes-4      915           889           -2.88%
BenchmarkChunkWriteQueue_addJob/1000_concurrent_writes-4     982           896           -8.76%
```

Reply:
I agree that it looks cleaner now, also moving the metric update outside of the lock is a great idea (btw, you can use a named `err error` return param and check it in the defer, so you wouldn't need the extra `success` variable)

I'd consider committing the benchmark to your test suite too.

Reply:
> I'd consider committing the benchmark to your test suite too

done: https://github.com/prometheus/prometheus/pull/10051/commits/ba8df032e5a9972771d9ee46a79ba47c6a7efbbd

> you can use a named err error return param and check it in the defer

done: https://github.com/prometheus/prometheus/pull/10051/commits/673040b0c56b1e81f0707b4531de8c19382fd004